lib/code_writer.rb

```diff
@@ -121,4 +102,37 @@ def write_arithmetic(command)
   def close
     @out.close
   end
+
+  private
+
+  def get_memory_address_for_pop(segment, index)
+    case segment
+    when *STATIC_SEGMENT_BASE_ADDRESS.keys
+      "@#{STATIC_SEGMENT_BASE_ADDRESS[segment] + index}"
+    when *DYNAMIC_SEGMENT_BASE_ADDRESS.keys
+       <<~EOF.chomp
+        @#{index}
+        D=A
+        @#{DYNAMIC_SEGMENT_BASE_ADDRESS[segment]}
+        A=M+D
+      EOF
+    end
+  end
+
+  def get_value_for_push(segment, index)
+    case segment
+    when *STATIC_SEGMENT_BASE_ADDRESS.keys
+      "@#{STATIC_SEGMENT_BASE_ADDRESS[segment] + index}\nD=M"
+    when *DYNAMIC_SEGMENT_BASE_ADDRESS.keys
+      <<~EOF.chomp
+        @#{index}
+        D=A
+        @#{DYNAMIC_SEGMENT_BASE_ADDRESS[segment]}
+        A=M+D
+        D=M
+      EOF
+    when "constant"
+      "@#{index}\nD=A"
```

Comment:
Fantastic. This change, plus the consistent shape of the two helper methods, makes an interesting fact way more obvious at a glance: `push` supports the `constant` segment (as a separate special case), but `pop` doesn’t.